proposals/module-types/Example-SharedEverythingDynamicLinking.md

```diff
@@ -0,0 +1,525 @@
+or not.
+
+Ultimately, the above versioning scheme may be too simplistic when many
+libraries and programs are used within a single application. Instead of simply
+embedding and propagating version numbers, semver *range patterns* may need to
+be specified for each dependency to provide developers finer-grained control
+over how the final version is chosen. Any such scheme would need to be tailored
+to a higher-level layer of tooling and should be able to built on the basic
+primitives of the Module Imports proposal, similar to the simplistic scheme
+shown above.
+
+## Cyclic Dependencies
+
+If cyclic dependencies are necessary, such cycles can be broken by:
+* identifying a [spanning] DAG over the module dependency graph;
+* keeping the calls along the spanning DAG's edges as normal function imports
+  and direct calls (as shown above); and
+* converting calls along "back edges" into indirect calls (`call_indirect`) of
+  an import const global `i32` index.
+
+For example, a cycle between modules `a` and `b` could be broken by having `b`
+directly import `a`:
+```wat
+;; a.wat
+(module
+  (type $Libc (instance ...))
+  (import "libc" (instance (type $Libc)))
+  (import "b" (instance $b (export "bar" (global i32 $bar))))
+  (type $FooType (func))
+  (func (export "foo")
+    (call_indirect $FooType (global.get $bar))
+  )
+)
+```
+```wat
+;; b.wat
+(module
+  (type $Libc (instance ...))
+  (import "libc" (module $LIBC (exports $Libc)))
+  (instance $libc (instance.instantiate $LIBC))
+
+  (module $INDICES (global $bar (export "bar") i32))
+  (instance $indices (instance.instantiate $INDICES))
+
+  (import "./a.wasm" (module $A
+    (import "libc" (instance (type $Libc)))
+    (import "indices" (instance (export "bar" (global i32))))
+    (export "foo" (func $foo))
+  ))
+  (instance $a (instance.instantiate $A (ref.instance $libc) (ref.instance $indices)))
+
+  ;; indirectly export bar to a:
+  (func $bar ...)
+  (elem (i32.const 0) $bar)
+  (func $start (global.set $indices.$bar (i32.const 0)))
+  (start $start)
+
+  (func (export "run")
+    call $a.$foo
+  )
+)
+```
+
+
+## Function Pointer Identity
+
+To ensure C function pointer identity across shared libraries, for each exported
+function, a shared library will need to export both the `func` definition and an
+`i32` `global` containing that `func`'s index in the global `funcref` table.
+
+Because a shared library can't know the absolute offset in the global `funcref`
+table for all of its exported functions, the table slots must be dynamically
+allocated. This can be achieved by the shared library calling into a `ftalloc`
+export of `libc` (analogous to `malloc`, but for allocating from the global
+`funcref` table) from the shared library's `start` function. Elements can
+then be written into the table (using [bulk memory operations]) at the allocated
+offset and their indices written into the exported `i32`s.
```

Comment:
Ah yes, good point; I'll soften the `must` wording here.

So I was imagining the `__table_base` strategy you describe when I wrote "(In theory, more efficient schemes are possible when the main program has more static knowledge of its shared libraries.)" below because, iiuc, for this to work the main module has to know statically how many slots the library needs before instantiating it.  (I suppose the main module could also probe for this dynamically (via `Module.customSections()` or something else), but the Module Linking proposal doesn't have the ability to do that from pure wasm.)

One thing I was concerned about is, at least with the static strategy, this would mean minor semver version updates could break main modules (in a rather silent way too).  A dynamic probing strategy could avoid this though.

Reply:
Yes, the current llvm+emscripten solution is dynamic and does involve a tiny custom section at the start of a shared library that specifies the number of slots it needs and the number of bytes of static data (along with alignment). 

I just realized that in theory the custom section could be avoided by looking that the segment lengths.. but then that wouldn't allow for bss / empty table slots.

Reply:
Cool, makes total sense.  In order to use Module Linking (which doesn't have the runtime ability to probe custom sections during instantiation), do you think it'd be possible to use the "a module allocates its own elem/data space" approach described here?

Reply:
Sure I see..   so in that case the static data payload would live in a passive segment and get loaded into a location returned from malloc?   I guess that works!   Then it can take that same address an store it in a private global that can be used to calculate load/store offsets (an non-exported internal version of `__memory_base`).

Reply:
Exactly, yes!  Cool.